File: BackEnd/src/agape/model/Fornecedor.java
package agape.model;

import agape.dao.FornecedorDAO;

import java.sql.Connection;
import java.util.List;

public class Fornecedor {
    private final FornecedorDAO dao = new FornecedorDAO();

    private int idFornec;
    private String nome;
    private String telefone1;
    private String telefone2;
    private String email;
    private String site;
    private String contato;
    private String cep;
    private String logradouro;
    private String cnpj;
    private String cidade;
    private String uf; /* SP RJ MG DF */
    private String obs;
    private int ativo;

    // GETTERS E SETTERS
    public int getIdFornec() {
        return idFornec;
    }
    public void setIdFornec(int idFornec) {
        this.idFornec = idFornec;
    }

    public String getNome() {
        return nome;
    }
    public void setNome(String nome) {
        this.nome = nome;
    }

    public String getTelefone1() {
        return telefone1;
    }
    public void setTelefone1(String telefone1) {
        this.telefone1 = telefone1;
    }

    public String getTelefone2() {
        return telefone2;
    }
    public void setTelefone2(String telefone2) {
        this.telefone2 = telefone2;
    }

    public String getEmail() {
        return email;
    }
    public void setEmail(String email) {
        this.email = email;
    }

    public String getSite() {
        return site;
    }
    public void setSite(String site) {
        this.site = site;
    }

    public String getContato() {
        return contato;
    }
    public void setContato(String contato) {
        this.contato = contato;
    }

    public String getCep() {
        return cep;
    }
    public void setCep(String cep) {
        this.cep = cep;
    }

    public String getLogradouro() {
        return logradouro;
    }
    public void setLogradouro(String logradouro) {
        this.logradouro = logradouro;
    }

    public String getCnpj() {
        return cnpj;
    }
    public void setCnpj(String cnpj) {
        this.cnpj = cnpj;
    }

    public String getCidade() {
        return cidade;
    }
    public void setCidade(String cidade) {
        this.cidade = cidade;
    }

    public String getUf() {
        return uf;
    }
    public void setUf(String uf) {
        this.uf = uf;
    }

    public String getObs() {
        return obs;
    }
    public void setObs(String obs) {
        this.obs = obs;
    }

    public int getAtivo() {
        return ativo;
    }
    public void setAtivo(int ativo) {
        this.ativo = ativo;
    }

    // ── Persistência (DAO encapsulado) ─────────────────────────────────────

    public List<Fornecedor> buscar(Connection conn, String nome, Integer ativo) throws Exception {
        return dao.buscar(conn, nome, ativo);
    }

    public Fornecedor buscarPorId(Connection conn, int id) throws Exception {
        return dao.buscarPorId(conn, id);
    }

    public boolean existeNome(Connection conn, String nome, int idExcluir) throws Exception {
        return dao.existeNome(conn, nome, idExcluir);
    }

    public void inserir(Connection conn) throws Exception {
        dao.inserir(conn, this);
    }

    public void atualizar(Connection conn) throws Exception {
        dao.atualizar(conn, this);
    }

    public void excluir(Connection conn) throws Exception {
        dao.excluir(conn, this.idFornec);
    }

    public String toJson() {
        return "{" +
            "\"idFornec\":"     + idFornec          + "," +
            "\"nome\":\""       + esc(nome)         + "\"," +
            "\"telefone1\":\""  + esc(telefone1)    + "\"," +
            "\"telefone2\":\""  + esc(telefone2)    + "\"," +
            "\"email\":\""      + esc(email)        + "\"," +
            "\"site\":\""       + esc(site)         + "\"," +
            "\"contato\":\""    + esc(contato)      + "\"," +
            "\"cep\":\""        + esc(cep)          + "\"," +
            "\"logradouro\":\"" + esc(logradouro)   + "\"," +
            "\"cnpj\":\""       + esc(cnpj)         + "\"," +
            "\"cidade\":\""     + esc(cidade)       + "\"," +
            "\"uf\":\""         + esc(uf)           + "\"," +
            "\"obs\":\""        + esc(obs)          + "\"," +
            "\"ativo\":"        + ativo             +
        "}";
    }

    private String esc(String s) {
        if (s == null) return "";
        return s.replace("\\", "\\\\").replace("\"", "\\\"");
    }
}

File: FrontEnd/models/Fornecedor.js
/* Fornecedor — entidade com validações e serialização para o backend */

window.AGAPE = window.AGAPE || {};
window.AGAPE.Models = window.AGAPE.Models || {};

window.AGAPE.Models.Fornecedor = (function () {

    function Fornecedor(dados) {
        dados = dados || {};
        this._idFornec   = dados.idFornec   || null;
        this._nome       = dados.nome       || '';
        this._telefone1  = dados.telefone1  || '';
        this._telefone2  = dados.telefone2  || '';
        this._email      = dados.email      || '';
        this._site       = dados.site       || '';
        this._contato    = dados.contato    || '';
        this._cep        = dados.cep        || '';
        this._logradouro = dados.logradouro || '';
        this._cnpj       = dados.cnpj       || '';
        this._cidade     = dados.cidade     || '';
        this._uf         = dados.uf         || '';
        this._obs        = dados.obs        || '';
        this._ativo      = _parsearAtivo(dados.ativo, 1);
    }

    function _parsearAtivo(valor, padrao) {
        if (valor === 1  || valor === '1'  || valor === true  || valor === 'true')  return 1;
        if (valor === 0  || valor === '0'  || valor === false || valor === 'false') return 0;
        return padrao;
    }

    Fornecedor.prototype.getIdFornec   = function () { return this._idFornec;   };
    Fornecedor.prototype.getNome       = function () { return this._nome;       };
    Fornecedor.prototype.getTelefone1  = function () { return this._telefone1;  };
    Fornecedor.prototype.getTelefone2  = function () { return this._telefone2;  };
    Fornecedor.prototype.getEmail      = function () { return this._email;      };
    Fornecedor.prototype.getSite       = function () { return this._site;       };
    Fornecedor.prototype.getContato    = function () { return this._contato;    };
    Fornecedor.prototype.getCep        = function () { return this._cep;        };
    Fornecedor.prototype.getLogradouro = function () { return this._logradouro; };
    Fornecedor.prototype.getCnpj       = function () { return this._cnpj;       };
    Fornecedor.prototype.getCidade     = function () { return this._cidade;     };
    Fornecedor.prototype.getUf         = function () { return this._uf;         };
    Fornecedor.prototype.getObs        = function () { return this._obs;        };
    Fornecedor.prototype.getAtivo      = function () { return this._ativo;      };

    Fornecedor.prototype.setIdFornec   = function (v) { this._idFornec   = v; };
    Fornecedor.prototype.setNome       = function (v) { this._nome       = v; };
    Fornecedor.prototype.setTelefone1  = function (v) { this._telefone1  = v; };
    Fornecedor.prototype.setTelefone2  = function (v) { this._telefone2  = v; };
    Fornecedor.prototype.setEmail      = function (v) { this._email      = v; };
    Fornecedor.prototype.setSite       = function (v) { this._site       = v; };
    Fornecedor.prototype.setContato    = function (v) { this._contato    = v; };
    Fornecedor.prototype.setCep        = function (v) { this._cep        = v; };
    Fornecedor.prototype.setLogradouro = function (v) { this._logradouro = v; };
    Fornecedor.prototype.setCnpj       = function (v) { this._cnpj       = v; };
    Fornecedor.prototype.setCidade     = function (v) { this._cidade     = v; };
    Fornecedor.prototype.setUf         = function (v) { this._uf         = v; };
    Fornecedor.prototype.setObs        = function (v) { this._obs        = v; };
    Fornecedor.prototype.setAtivo      = function (v) { this._ativo      = _parsearAtivo(v, 1); };

    Fornecedor.prototype.validarNome = function () {
        return typeof this._nome === 'string' && this._nome.trim().length >= 2;
    };

    Fornecedor.prototype.validar = function () {
        var erros = [];
        if (!this.validarNome()) erros.push('Nome deve ter pelo menos 2 caracteres.');
        return erros;
    };

    Fornecedor.prototype.paraFormData = function () {
        var dados = {
            nome:       this._nome,
            telefone1:  this._telefone1,
            telefone2:  this._telefone2,
            email:      this._email,
            site:       this._site,
            contato:    this._contato,
            cep:        this._cep,
            logradouro: this._logradouro,
            cnpj:       this._cnpj,
            cidade:     this._cidade,
            uf:         this._uf,
            obs:        this._obs,
            ativo:      this._ativo
        };
        if (this._idFornec !== null && this._idFornec !== undefined) {
            dados.idFornec = this._idFornec;
        }
        return dados;
    };

    return Fornecedor;

})();
